eng/packages/visitors/Visitors/src/Shared/SpecialHeadersVisitor.cs

```diff
@@ -66,28 +64,35 @@ internal class SpecialHeadersVisitor : ScmLibraryVisitor
                         })
                     {
                         var variable = declaration.Variable;
-                        if (variable.Type.Equals(typeof(Request)))
+                        if (variable.Type.Equals(variable.ToApi<HttpRequestApi>().Type))
                         {
                             requestVariable = variable;
                         }
                     }
                 }
 
-                if (clientRequestIdParameter != null)
+                // Only set the client-request-id header if the request type supports it
+                string? requestIdPropName = requestVariable?.ToApi<HttpRequestApi>().HttpRequestClientIdPropertyName;
+                if (requestIdPropName != null && clientRequestIdParameter != null)
                 {
                     // Set the client-request-id header
-                    newStatements.Add(requestVariable!.As<Request>().SetHeaderValue(
-                        clientRequestIdParameter.NameInRequest, requestVariable.Property(nameof(Request.ClientRequestId))));
+                    newStatements.Add(requestVariable!.ToApi<HttpRequestApi>().SetHeaders(
+                    [
+                        Literal(clientRequestIdParameter.NameInRequest),
+                        requestVariable!.Property(requestIdPropName)
```

Comment:
we should not be using the abstractions this way.
We should do this instead:
```suggestion
                        requestVariable!.ToApi<HttpRequestApi>().ClientRequestId()
```